### PR TITLE
Latest Esprima beta changes

### PR DESCRIPTION
**Pull request by AmateurMemetics**
This update is a pre-requisite for adding arrow functions.

Unfortunately as I'm not completely familiar with the code / intentions behind all of the changes in Esprima, I may have gone about this in the wrong way (in particular I do recall reading that `Esprima.Ast.List` is not supposed to be exposed?).

I'd be happy to take any comments on board. Once this is in, the change to add arrow functions is very minor.

### CODE REVIEW

File: Jint/Engine.cs [hunk 1]
```diff
@@ -762,8 +762,8 @@ public JsValue GetValue(JsValue scope, string propertyName)
         //  http://www.ecma-international.org/ecma-262/5.1/#sec-10.5
         internal bool DeclarationBindingInstantiation(
             DeclarationBindingType declarationBindingType,
-            List<FunctionDeclaration> functionDeclarations,
-            List<VariableDeclaration> variableDeclarations,
+            Esprima.Ast.List<FunctionDeclaration> functionDeclarations,
```
Comment (lahma) [inline Jint/Engine.cs]: these should be passed as refs as Esprima's list is a struct

File: Jint/Engine.cs [hunk 2]
```diff
@@ -847,7 +847,7 @@ public JsValue GetValue(JsValue scope, string propertyName)
             return canReleaseArgumentsInstance;
         }
 
-        private void AddFunctionDeclarations(List<FunctionDeclaration> functionDeclarations, EnvironmentRecord env, bool configurableBindings, bool strict)
+        private void AddFunctionDeclarations(Esprima.Ast.List<FunctionDeclaration> functionDeclarations, EnvironmentRecord env, bool configurableBindings, bool strict)
```
Comment (lahma) [inline Jint/Engine.cs]: this should be passed as ref as Esprima's list is a struct

File: Jint/Runtime/Environments/DeclarativeEnvironmentRecord.cs [hunk 3]
```diff
@@ -384,7 +384,7 @@ internal static JsValue HandleAssignmentPatternIfNeeded(IFunction functionDeclar
             return defaultValue;
         }
 
-        internal void AddVariableDeclarations(List<VariableDeclaration> variableDeclarations)
+        internal void AddVariableDeclarations(Esprima.Ast.List<VariableDeclaration> variableDeclarations)
```
Comment (lahma) [inline Jint/Runtime/Environments/DeclarativeEnvironmentRecord.cs]: this should be passed as ref as Esprima's list is a struct

File: Jint/Runtime/Interpreter/JintStatementList.cs [hunk 4]
```diff
@@ -15,12 +15,12 @@ private class Pair
...
         private readonly Engine _engine;
         private readonly Statement _statement;
-        private readonly List<StatementListItem> _statements;
+        private readonly Esprima.Ast.List<StatementListItem> _statements;
 
         private Pair[] _jintStatements;
         private bool _initialized;
 
-        public JintStatementList(Engine engine, Statement statement, List<StatementListItem> statements)
+        public JintStatementList(Engine engine, Statement statement, Esprima.Ast.List<StatementListItem> statements)
```
Comment (lahma) [inline Jint/Runtime/Interpreter/JintStatementList.cs]: statements probably should be a ref param

File: Jint/Runtime/Interpreter/Statements/JintSwitchBlock.cs [hunk 5]
```diff
@@ -8,11 +9,11 @@ namespace Jint.Runtime.Interpreter.Statements
     internal sealed class JintSwitchBlock
     {
         private readonly Engine _engine;
-        private readonly List<SwitchCase> _switchBlock;
+        private readonly Esprima.Ast.List<SwitchCase> _switchBlock;
         private JintSwitchCase[] _jintSwitchBlock;
         private bool _initialized;
 
-        public JintSwitchBlock(Engine engine, List<SwitchCase> switchBlock)
+        public JintSwitchBlock(Engine engine, Esprima.Ast.List<SwitchCase> switchBlock)
```
Comment (lahma) [inline Jint/Runtime/Interpreter/Statements/JintSwitchBlock.cs]: switchBlock probably should be a ref param

**Comment by Happypig375** [comment]
I think using the `in` modifier instead of `ref` would be better since it guarantees the list cannot be reassigned.

**Comment by lahma** [comment]
But as Esprima list is not read-only I believe a defensive copy would be made.

**Comment by AmateurMemetics** [comment]
I agree with @lahma. I've just been reading this blog post to get a better grasp on the `in` modifier in C#: https://blogs.msdn.microsoft.com/seteplia/2018/03/07/the-in-modifier-and-the-readonly-structs-in-c/ - which concludes that a defensive copy will be made for structs which are not read-only. 

I'll go ahead and update this PR with `ref` keywords.

File: Jint/Runtime/Interpreter/Statements/JintSwitchBlock.cs [hunk 6]
```diff
@@ -8,11 +9,11 @@ namespace Jint.Runtime.Interpreter.Statements
     internal sealed class JintSwitchBlock
     {
         private readonly Engine _engine;
-        private readonly List<SwitchCase> _switchBlock;
+        private readonly Esprima.Ast.List<SwitchCase> _switchBlock;
         private JintSwitchCase[] _jintSwitchBlock;
         private bool _initialized;
 
-        public JintSwitchBlock(Engine engine, List<SwitchCase> switchBlock)
+        public JintSwitchBlock(Engine engine, in Esprima.Ast.List<SwitchCase> switchBlock)
```
Comment (AmateurMemetics) [inline Jint/Runtime/Interpreter/Statements/JintSwitchBlock.cs]: Used an `in` modifier here because `_switchBlock` is `readonly`

**Comment by AmateurMemetics** [comment]
Is there anything stopping this from being merged in?

**Comment by lahma** [comment]
Sorry to keep you waiting. @sebastienros makes the final call and I believe he's been a bit under the weather, hopefully back soon to review.

**Pull request merged**